File: src/crypto.c
/*==============================================================================
 *     File: crypto.c
 *  Created: 10/19/2016, 22:17
 *   Author: Bernie Roesler
 *
 *  Description: Utility functions for cryptography challenges
 *
 *============================================================================*/
#include "crypto.h"
#include "crypto_util.h"
#include "header.h"

/*------------------------------------------------------------------------------
 *      Convert hexadecimal string to base64 string
 *----------------------------------------------------------------------------*/
char *hex2b64_str(char *hex_str)
{
    const char *b64_lut = 
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=";
    int nchr_in,
        nbyte_in,
        nbyte_out,
        nchr_out,
        b64_int;
    int hex_int;

    if (hex_str) {
        nchr_in = strlen(hex_str);      /* Number of chars in encoded string */
    } else {
        return NULL;
    }

    if (nchr_in & 1) {
        ERROR("Input string it not a valid hex string!");
    } else {
        nbyte_in = nchr_in / 2;            /* 2 hex chars == 1 byte */
    }

    /* round nbyte up to multiple of 3 (need 3 bytes for even base64 output) */
    nbyte_out = nbyte_in + 3 - (nbyte_in % 3);
    nchr_out = nbyte_out * 4/3;  /* Number of chars in output */

    /* allocate memory for output */
    char *b64_str = malloc(nchr_out * sizeof(char));
    MALLOC_CHECK(b64_str);
    BZERO(b64_str, nchr_out);

    /* Operate in chunks of 3 bytes in ==> 4 bytes out */
    for (int i = 0; i < nbyte_in; i+=3) {
        int j = 0;
        hex_int = getHexByte(hex_str+2*i+2*j);

        /* Add first character using first 6 bits of first byte */
        b64_int = (hex_int & 0xFC) >> 2;
        strncat(b64_str, &b64_lut[b64_int], 1);

        /* get last 2 bits of first byte */
        b64_int = (hex_int & 0x03) << 4;

        /* if we have more bytes to go */
        if (j+1 < nbyte_in) {
            j++;
            hex_int = getHexByte(hex_str+2*i+2*j);

            /* Add second character using first 4 bits of second byte and
             * combine with 2 from above */
            b64_int |= (hex_int & 0xF0) >> 4;
            strncat(b64_str, &b64_lut[b64_int], 1);

            /* get last 4 bits of second byte */
            b64_int = (hex_int & 0x0F) << 2;

            /* if we have more bytes to go */
            if (j+1 < nbyte_in) {
                j++;
                hex_int = getHexByte(hex_str+2*i+2*j);
                /* Add third character */
                /* get first 2 bits of third byte and combine with 4 from above */
                b64_int |= (hex_int & 0xC0) >> 6;
                strncat(b64_str, &b64_lut[b64_int], 1);

                /* Add fourth character using last 6 bits of third byte */
                b64_int = (hex_int & 0x3F);
                strncat(b64_str, &b64_lut[b64_int], 1);

            /* There are only 2 bytes of input, so interpret 3rd character with
             * a "0x00" byte appended, and pad with an '=' character */
            } else {
                strncat(b64_str, &b64_lut[b64_int], 1);
                strncat(b64_str, "=", 1);
            }

        /* There is only 1 byte of input, so interpret 2nd character with two
         * "0x00" bytes appended, and pad with an '=' character */
        } else {
            strncat(b64_str, &b64_lut[b64_int], 1);
            strncat(b64_str, "==", 2);
        }
    }

    return b64_str;
}

/*------------------------------------------------------------------------------
 *      XOR two equal-length hex-encoded buffers
 *----------------------------------------------------------------------------*/
char *fixedXOR(char *str1, char *str2)
{
    size_t len1 = strlen(str1),
           len2 = strlen(str2);
    char *hex_str;
    int hex_xor, hex_int1, hex_int2;
    char hex_chars[3];

    if (len1 != len2) { ERROR("Input strings must be the same length!"); }

    /* int nbyte = len1/2; */

    /* allocate memory for string output */
    hex_str = malloc(len1 * sizeof(char));
    MALLOC_CHECK(hex_str);
    BZERO(hex_str, len1);

    /* XOR each byte in the input string */
    for (int i = 0; i < len1; i+=2) {
        hex_int1 = getHexByte(str1+i); /* 2 chars per byte */
        hex_int2 = getHexByte(str2+i);
        hex_xor = hex_int1 ^ hex_int2;
        snprintf(hex_chars, 3, "%0.2X", hex_xor);  /* convert to hex chars */
        strncat(hex_str, hex_chars, 2);            /* append to output string */
    }

    /* return hex-encoded string */
    return hex_str;
}

/*------------------------------------------------------------------------------
 *         Find character frequency in string 
 *----------------------------------------------------------------------------*/
CHARFREQ *countChars(char *s)
{
    /* TODO update this function to count other non-alphabetic characters, and
     * penalize strings that have many non-ascii characters */
    CHARFREQ *cf; /* one struct per letter of alphabet */
    int i;

    /* Initialize struct array */
    cf = malloc(NUM_LETTERS * sizeof(CHARFREQ));
    MALLOC_CHECK(cf);
    BZERO(cf, NUM_LETTERS*sizeof(CHARFREQ));

    /* Populate struct array with ALL characters */
    for (i = 0; i < NUM_LETTERS; i++) {
        cf[i].letter = i;
        cf[i].count = 0;
    }

    /* Get frequency of ALL characters in the string */
    i = 0;
    while (s[i]) {
        cf[(size_t)s[i]].count++;
        i++;
    }
    return cf;
}

/*------------------------------------------------------------------------------
 *         Get character frequency score of string
 *----------------------------------------------------------------------------*/
float charFreqScore(char *str)
{
    const char etaoin[] = "ETAOINSHRDLCUMWFGYPBVKJXQZ";  /* acceptable chars */
    int len = strlen(etaoin);
    float observed = 0.0,
          expected = 0.0,
          chi_sq = 0.0;
    int ch_ind;

    /* Get ordered string of letters */
    CHARFREQ *cf = countChars(str);

    /* <https://en.wikipedia.org/wiki/Letter_frequency> */
    const float english_freq[] = 
        { 0.08167, 0.01492, 0.02782, 0.04253, 0.12702, 0.02228, 0.02015,  \
          0.06094, 0.06966, 0.00153, 0.00772, 0.04025, 0.02406, 0.06749,  \
          0.07507, 0.01929, 0.00095, 0.05987, 0.06327, 0.09056, 0.02758,  \
          0.00978, 0.02360, 0.00150, 0.01974, 0.00074 };

    /* Calculate score via chi-squared test */
    float slen = (float)strlen(str);

    /* Sum the chi^2 values for each alphabetic character */
    for (int i = 0; i < len; i++) {
        ch_ind = etaoin[i];
        observed = cf[ch_ind].count;
        expected = english_freq[ch_ind-'A'] * slen;
        chi_sq += (observed - expected)*(observed - expected) / expected;
    }

    free(cf);
    return chi_sq;
}

/*------------------------------------------------------------------------------
 *         Encode a string with a single byte XOR cipher 
 *----------------------------------------------------------------------------*/
char *singleByteXOREncode(char *hex, int key_int)
{
    size_t len = strlen(hex);
    if (len & 1) { ERROR("Input string is not a valid hex string!"); }
    if ((key_int < 0x00) || (key_int >= 0x100)) { ERROR("key is outside of valid range!"); }

    int nbyte = len/2;
    char key[3],            /* i.e. 0x01 --> '01' */
         key_str[len+1];    /* i.e. if hex == "4D616E", key_str = "010101" */
    BZERO(key, 3);
    BZERO(key_str, len+1);

    /* repeat key for each byte of input, so only one XOR is needed */
    snprintf(key, 3, "%0.2X", key_int);
    for (int j = 0; j < nbyte; j++) {
        strncat(key_str, key, 2);
    }

    /* XOR each byte in the ciphertext with the key */
    return fixedXOR(hex, key_str);
}

/*------------------------------------------------------------------------------
 *         Decode a string XOR'd against a single character
 *----------------------------------------------------------------------------*/
char *singleByteXORDecode(char *hex)
{
    size_t len = strlen(hex);
    if (len & 1) { ERROR("Input string is not a valid hex string!"); }

    float cfreq_score_min = 1.0e9; /* initialize to large number */

    /* Allocate memory for the output */
    char *plaintext = malloc(len * sizeof(char));
    MALLOC_CHECK(plaintext);
    BZERO(plaintext, len);

    char *true_key = malloc(3 * sizeof(char));
    MALLOC_CHECK(true_key);
    BZERO(true_key, sizeof(*true_key));

    /* test each possible character byte */
#ifdef LOGSTATUS
        printf("key\tdecoded string\t\t\t\tchi^2\n");
#endif
    /* for (int i = 0x58; i < 0x59; i++) { */
    for (int i = 0x00; i < 0x100; i++) {
        /* Decode hex string */
        char *xor = singleByteXOREncode(hex, i);

        /* Convert to plain ASCII text */
        char *ptext = htoa(xor);

        /* Calculate character frequency score */
        float cfreq_score = charFreqScore(ptext);

#ifdef LOGSTATUS
        /* print each key, decoded string, score */
        if (isValid(ptext) == 0) {
            printf("%0.2X\t%s\t%10.4f\n", i, ptext, cfreq_score);
        } else {
            printf("%0.2X\t%s\t\t%10.5f\n", i, "--------------------", cfreq_score);
        }
#endif

        /* Track minimum chi-squared score and actual key */
        if (cfreq_score < cfreq_score_min) {
            cfreq_score_min = cfreq_score;
            snprintf(true_key, 3, "%0.2X", i);
            strncpy(plaintext, ptext, len);
        }
    }

#ifdef LOGSTATUS
    printf("--------------------\n");
    printf("found key = %s\n", true_key);
    printf("cfreq_score_min = %10.4f\n", cfreq_score_min);
#endif

    /* Return the decoded plaintext! */
    return plaintext;
}
/*==============================================================================
 *============================================================================*/

File: python/crypto.py
#!/usr/bin/python
#==============================================================================
#     File: crypto.py
#  Created: 03/03/2016, 14:55
#   Author: Bernie Roesler
#
'''
  Functions to support solutions to Matasano Crypto Challenges, Set 1.
'''
#==============================================================================
from collections import namedtuple
import string
# import pdb
# import traceback
# import sys

# base64 look-up table
b64_lut = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/='

#------------------------------------------------------------------------------
#       Convert base64 string to hexadecimal string
#------------------------------------------------------------------------------
def b642hex_str(b64_str):
    '''Convert base64 string to hex string.'''
    nbyte = len(b64_str)     # Number of chars in encoded string == # bytes

    if nbyte % 4 != 0:
        raise ValueError('Input string must have a multiple of 4 characters!')
        
    # List of integers corresponding to b64 characters in input
    b64_byte = [b64_lut.find(b64_str[i]) for i in range(0, nbyte)]

    hex_int = []
    for i in range(0, nbyte, 4):
        # Take chunks of 4 bytes --> 3 bytes of output
        chunk = b64_byte[i:i+4]

        # First char of output
        #   Need to mask off MSBs for left-shifts so we don't keep large #s
        hex_int.append((chunk[0] << 2) & 0xFF | (chunk[1] >> 4))

        # Second char
        if (chunk[2] < 0x40) and (chunk[2] > 0x00):
            hex_int.append((chunk[1] << 4) & 0xFF | (chunk[2] >> 2))

            # Third char
            if (chunk[3] < 0x40) and (chunk[3] > 0x00):
               hex_int.append((chunk[2] << 6) & 0xFF | chunk[3])

    # Convert integer list to hex string
    hex_str = ''.join(['%02x'%k for k in hex_int])

    return hex_str

#------------------------------------------------------------------------------
#       Convert hexadecimal string to base64 string
#------------------------------------------------------------------------------
def hex2b64_str(hex_str):
    '''Convert hex string to base64 string.'''

    nchr_in = len(hex_str)     # Number of chars in encoded string
    nbyte = nchr_in / 2         # 2 hex chars == 1 byte
    # nchr_out = nbyte * 4/3      # Number of chars in output

    # Byte array (2 hex chars == 1 byte)
    hex_byte = [int(hex_str[i:i+2], 16) for i in range(0, nchr_in, 2)]

    b64_str = ''
    # Operate in chunks of 3 bytes  in ==> 4 bytes out
    for i in range(0, nbyte, 3):
        # Add first character using first 6 bits of first byte
        # Need 2 chars of hex to get 1 byte
        b64_int = (hex_byte[i] & 0xFC) >> 2
        b64_str += b64_lut[b64_int]

        # get last 2 bits of first byte
        b64_int = (hex_byte[i] & 0x03) << 4

        # if we have more bytes to go
        if i+1 < nbyte:
            # Add second character using first 4 bits of second byte and
            # combine with 2 from above
            b64_int |= (hex_byte[i+1] & 0xF0) >> 4
            b64_str += b64_lut[b64_int]

            # get last 4 bits of second byte
            b64_int = (hex_byte[i+1] & 0x0F) << 2

            # if we have more bytes to go
            if i+2 < nbyte:
                # Add third character
                # get first 2 bits of third byte and combine with 4 from above
                b64_int |= (hex_byte[i+2] & 0xC0) >> 6

                # # The following construct is the same as "dbstop if error":
                # try:
                #     b64_int |= (hex_byte[i+2] & 0xC0) >> 6
                # except:
                #     type, value, tb = sys.exc_info()
                #     traceback.print_exc()
                #     pdb.post_mortem(tb)

                b64_str += b64_lut[b64_int]

                # Add fourth character using last 6 bits of third byte
                b64_int = (hex_byte[i+2] & 0x3F)
                b64_str += b64_lut[b64_int]

            # There are only 2 bytes of input, so interpret 3rd character with
            # a "0x00" byte appended, and pad with an '=' character
            else:
                b64_str += b64_lut[b64_int]
                b64_str += '='

        # There is only 1 byte of input, so interpret 2nd character with
        # two "0x00" bytes appended, and pad with an '=' character
        else:
            b64_str += b64_lut[b64_int]
            b64_str += '=='

    return b64_str

#------------------------------------------------------------------------------
#       Fixed-length XOR
#------------------------------------------------------------------------------
def fixedXOR(str1, key):
    '''Take a hex-encoded string and XOR it with a hex-encoded key to return
    an integer.'''

    if len(str1) != len(key):
        raise ValueError('Input strings must be of equal length!')

    # XOR the numbers
    out_int = int(str1, 16) ^ int(key, 16)

    return out_int

#------------------------------------------------------------------------------
#      Character frequency score (English)
#------------------------------------------------------------------------------
# def char_freq_score(plaintext):
#     '''Score an English string on a scale of 0 to 12 based on character
#     frequency.'''
#     # Set of printable characters
#     valid_characters = set(string.printable)-set(string.digits)-set(['#','$','%','/','~','`'])
#
#     if all(char in valid_characters for char in plaintext):
#         score = 100
#     else:
#         score = 0
#
#     return score

def char_freq_score(plaintext):
    '''Score an English string on a scale of 0 to 12 based on character
    frequency.'''
    # Most common English letters in order (include space!!)
    etaoin = ' etaoinshrdlcumwfgypbvkjxqz'

    # Get list of lowercase letters in order of most frequency to least
    freqOrder = get_frequency_order(plaintext)

    # Find matches in top N characters
    N = 22
    score = 0
    for ch in etaoin[:N]:
        if ch in freqOrder[:N]:
            score += 1

    return score

#------------------------------------------------------------------------------
#       String character frequency order
#------------------------------------------------------------------------------
def get_frequency_order(plaintext):
    '''Return string of character frequencies in input string ranked highest
    to lowest.'''

    # Build dictionary
    mydict = {}
    for c in plaintext:
        if c not in mydict:
            mydict[c] = 1
        else:
            mydict[c] += 1

    # Sort by character frequency, return string instead of list
    ranks = ''.join(sorted(mydict, key=mydict.get, reverse=True))

    return ranks

#------------------------------------------------------------------------------
#       Decode XOR input with single byte
#------------------------------------------------------------------------------
def single_byte_XOR_decode(ciphertext):
    '''
    Take a hex-encoded string that has been XOR'd against a single
    character, and decode it.
    Input:
        ciphertext      hex-encoded string
    Output:
        out.key         integer value of the "true key" used for decryption
        out.decrypt     actual decrypted string
        out.score       integer character frequency score
    '''
    # Input checking
    N = len(ciphertext)   # number of bytes in ciphertext

    if N % 2 != 0:
        raise ValueError('Input string must have even number of characters!')

    # Initialize variables
    cfreq_score_max = 0
    true_key = 0x00
    plaintext_decrypt = ''

    # For each possible byte
    for i in range(0x00, 0x100):
        # XOR each character in ciphertext with key
        key = hex(i).lstrip('0x') or '0'

        # Single-digit hex numbers need a padding 0
        if len(key) == 1:
            key = '0' + key

        plaintext = ''
        # XOR each byte in the ciphertext with the key (1 byte == 2 hex chars)
        for cipherbyte in [ciphertext[j:j+2] for j in range(0, N, 2)]:
            plaintext += chr(fixedXOR(cipherbyte, key))

        # Calculate character frequency score
        cfreq_score = char_freq_score(plaintext)

        # Track maximum score and actual key
        if cfreq_score >= cfreq_score_max:
            cfreq_score_max   = cfreq_score # int
            true_key          = key         # chr (in hex!)
            plaintext_decrypt = plaintext   # str

    # Store output in named tuple
    output = namedtuple('output', 'key decrypt score')
    out = output(true_key, plaintext_decrypt, cfreq_score_max)

    return out

#------------------------------------------------------------------------------
#       Repeating key XOR
#------------------------------------------------------------------------------
def repeating_key_XOR(plaintext_hex, key_hex):
    '''Return an encrypted string from a hex-encoded input, XORed with
    a repeating key.'''

    N = len(plaintext_hex)
    M = len(key_hex)

    pt_byte = [plaintext_hex[i:i+2] for i in range(0, N, 2)]
    key_byte = [key_hex[i:i+2] for i in range(0, M, 2)]

    ciphertext_hex = ''
    for i in range(0, N/2):
        ciphertext_int = fixedXOR(pt_byte[i], key_byte[i % (M/2)])
        ciphertext_hex += '{0:02x}'.format(ciphertext_int)

    return ciphertext_hex

#------------------------------------------------------------------------------
#       Hamming distance
#------------------------------------------------------------------------------
def hamming_dist(str1, str2):
    '''Returns the number of differing bits of two hex-encoded strings.'''
    # XOR the strings to get 1's where the bits differ
    str_xor = fixedXOR(str1,str2)

    # Count the number of bits that differ
    H = bin(str_xor).count('1')

    return H
#==============================================================================
#==============================================================================
